File: apps/python-sdk/firecrawl/v2/methods/map.py
"""
Mapping functionality for Firecrawl v2 API.
"""

from typing import Optional, Dict, Any
from ..types import MapOptions, MapData, LinkResult
from ..utils import HttpClient, handle_response_error


def _prepare_map_request(url: str, options: Optional[MapOptions] = None) -> Dict[str, Any]:
    if not url or not url.strip():
        raise ValueError("URL cannot be empty")

    payload: Dict[str, Any] = {"url": url.strip()}

    if options is not None:
        # Unified sitemap parameter already provided in options
        data: Dict[str, Any] = {}
        if getattr(options, "sitemap", None) is not None:
            data["sitemap"] = options.sitemap

        if options.search is not None:
            data["search"] = options.search
        if options.include_subdomains is not None:
            data["includeSubdomains"] = options.include_subdomains
        if options.limit is not None:
            data["limit"] = options.limit
        if options.timeout is not None:
            data["timeout"] = options.timeout
        payload.update(data)

    return payload


def map(client: HttpClient, url: str, options: Optional[MapOptions] = None) -> MapData:
    """
    Map a URL and return MapData (links list with optional titles/descriptions).
    """
    request_data = _prepare_map_request(url, options)
    response = client.post("/v2/map", request_data)
    if not response.ok:
        handle_response_error(response, "map")

    body = response.json()
    if not body.get("success"):
        raise Exception(body.get("error", "Unknown error occurred"))

    # shouldnt return inside data?
    # data = body.get("data", {})
    # result_links: list[LinkResult] = []
    # for item in data.get("links", []):
    #     if isinstance(item, dict):
    #         result_links.append(
    #             LinkResult(
    #                 url=item.get("url", ""),
    #                 title=item.get("title"),
    #                 description=item.get("description"),
    #             )
    #         )
    #     elif isinstance(item, str):
    #         result_links.append(LinkResult(url=item))

    result_links: list[LinkResult] = []
    for item in body.get("links", []):
        if isinstance(item, dict):
            result_links.append(
                LinkResult(
                    url=item.get("url", ""),
                    title=item.get("title"),
                    description=item.get("description"),
                )
            )
        elif isinstance(item, str):
            result_links.append(LinkResult(url=item))

    return MapData(links=result_links)

File: apps/js-sdk/firecrawl/src/v2/methods/map.ts
import { type MapData, type MapOptions, type SearchResult } from "../types";
import { HttpClient } from "../utils/httpClient";
import { throwForBadResponse, normalizeAxiosError } from "../utils/errorHandler";

function prepareMapPayload(url: string, options?: MapOptions): Record<string, unknown> {
  if (!url || !url.trim()) throw new Error("URL cannot be empty");
  const payload: Record<string, unknown> = { url: url.trim() };
  if (options) {
    if (options.sitemap != null) payload.sitemap = options.sitemap;
    if (options.search != null) payload.search = options.search;
    if (options.includeSubdomains != null) payload.includeSubdomains = options.includeSubdomains;
    if (options.limit != null) payload.limit = options.limit;
    if (options.timeout != null) payload.timeout = options.timeout;
  }
  return payload;
}

export async function map(http: HttpClient, url: string, options?: MapOptions): Promise<MapData> {
  const payload = prepareMapPayload(url, options);
  try {
    const res = await http.post<{ success: boolean; error?: string; links?: Array<string | SearchResult> }>("/v2/map", payload);
    if (res.status !== 200 || !res.data?.success) {
      throwForBadResponse(res, "map");
    }
    const linksIn = res.data.links || [];
    const links: SearchResult[] = [];
    for (const item of linksIn) {
      if (typeof item === "string") links.push({ url: item });
      else if (item && typeof item === "object") links.push({ url: item.url, title: (item as any).title, description: (item as any).description });
    }
    return { links };
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "map");
    throw err;
  }
}

